Lib/asyncio/tasks.py

```diff
@@ -442,62 +443,12 @@ async def wait_for(fut, timeout):
 
     This function is a coroutine.
     """
-    loop = events.get_running_loop()
-
-    if timeout is None:
-        return await fut
-
-    if timeout <= 0:
-        fut = ensure_future(fut, loop=loop)
-
-        if fut.done():
-            return fut.result()
-
-        await _cancel_and_wait(fut, loop=loop)
-        try:
-            return fut.result()
-        except exceptions.CancelledError as exc:
-            raise exceptions.TimeoutError() from exc
-
-    waiter = loop.create_future()
-    timeout_handle = loop.call_later(timeout, _release_waiter, waiter)
-    cb = functools.partial(_release_waiter, waiter)
 
-    fut = ensure_future(fut, loop=loop)
-    fut.add_done_callback(cb)
+    async def inner():
```

Comment:
@asvetlov I've been having a go at this branch in bf594bd5a307660d36ab627acee80e34777e3820

you need to special case `None` in `asyncio.wait_for(`

```suggestion
    if timeout is None:
        return await fut

    async def inner():
```

and then special case `0` in timeouts.timeout see https://gist.github.com/graingert/ea2546b23b32be5a4493a9a115db2eff#file-timeout_bug-py-L6